components/style/stylesheets.rs

```diff
@@ -76,21 +70,24 @@ impl From<SingleRuleParseError> for RulesMutateError {
     }
 }
 
+fn only_ns_or_import(rules: &[CssRule]) -> bool {
+    rules.iter().all(|r| {
+        match *r {
+            CssRule::Namespace(..) /* | CssRule::Import(..) */ => true,
+            _ => false
+        }
+    })
+}
+
 impl CssRules {
-    // used in CSSOM
-    pub fn only_ns_or_import(rules: &[CssRule]) -> bool {
-        rules.iter().all(|r| {
-            match *r {
-                CssRule::Namespace(..) /* | CssRule::Import(..) */ => true,
-                _ => false
-            }
-        })
+    pub fn new(rules: Vec<CssRule>) -> Arc<RwLock<CssRules>> {
+        Arc::new(RwLock::new(CssRules(rules)))
     }
 
     // https://drafts.csswg.org/cssom/#insert-a-css-rule
-    pub fn insert_rule(&self, rule: &str, base_url: ServoUrl, index: usize, nested: bool)
+    pub fn insert_rule(&mut self, rule: &str, base_url: ServoUrl, index: usize, nested: bool)
```

Comment:
This extends the lifetime of the write lock for much longer. I don't like it; there's an arbitrarily complex parse step in the critical section.

Actually, looks like that already happened in your previous change to move the `insert_rule` logic into this crate -- when it was in the dom crate there was a separate read lock followed by a short-lived write lock. However, in that change it is still possible to use a read lock followed by a write lock, it's just that we don't do that. After this change it is no longer possible.